source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2696 @@
+      - session: { id: session1, client: session0 }
+
+
+Supported Entity Types
+~~~~~~~~~~~~~~~~~~~~~~
+
+Test runners MUST support the following types of entities:
+
+- MongoClient. See `entity_client`_ and `client`_.
+- Database. See `entity_database`_ and `database`_.
+- Collection. See `entity_collection`_ and `collection`_
+- ClientSession. See `entity_session`_ and `session`_.
+- GridFS Bucket. See `entity_bucket`_ and `bucket`_.
+- GridFS Stream. See `entity_stream`_.
+- ChangeStream. See `changeStream`_.
+- The following BSON types and/or equivalent language types:
+
+  - 0x01 - 0x13
+  - 0xFF
+  - 0x7F
+
+This is an exhaustive list of permissible types for the entity map. Test runners
+MUST raise an error if another type would be added to the entity map.
+
+
+Test Format
+-----------
+
+Each specification test file can define one or more tests, which inherit some
+top-level configuration (e.g. namespace, initial data). YAML and JSON test files
+are parsed as an object by the test runner. This section defines the top-level
+keys for that object and links to various sub-sections for definitions of
+nested structures (e.g. individual `test`_, `operation`_).
+
+Although test runners are free to process YAML or JSON files, YAML is the
+canonical format for writing tests. YAML files may be converted to JSON using a
+tool such as `js-yaml <https://github.com/nodeca/js-yaml>`__ .
+
+
+Top-level Fields
+~~~~~~~~~~~~~~~~
+
+The top-level fields of a test file are as follows:
+
+- ``description``: Required string. The name of the test file.
+
+  This SHOULD describe the common purpose of tests in this file and MAY refer to
+  the filename (e.g. "updateOne-hint").
+
+.. _schemaVersion:
+
+- ``schemaVersion``: Required string. Version of this specification to which the
+  test file complies.
+
+  Test files SHOULD be conservative when specifying a schema version. For
+  example, if the latest schema version is 1.1 but the test file complies with
+  schema version 1.0, the test file should specify 1.0.
+
+  Test runners will use this to determine compatibility (i.e. whether and how
+  the test file will be interpreted). The format of this string is defined in
+  `Version String`_; however, test files SHOULD NOT need to refer to specific
+  patch versions since patch-level changes SHOULD NOT alter the structure of the
+  test format (as previously noted in `Schema Version`_).
+
+.. _runOnRequirements:
+
+- ``runOnRequirements``: Optional array of one or more `runOnRequirement`_
+  objects. List of server version and/or topology requirements for which the
+  tests in this file can be run. If no requirements are met, the test runner
+  MUST skip this test file.
+
+.. _createEntities:
+
+- ``createEntities``: Optional array of `entity`_ objects. List of entities
+  (e.g. client, collection, session objects) that should be created before each
+  test case is executed.
+
+  Test files SHOULD define entities in dependency order, such that all
```

Comment:
This isn't something we can enforce in the schema. If a test file defines things out of order, I would expect the test runner to fail. I would plan to write such a test to assert test runner behavior.

Would it be helpful to clarify why I'm using SHOULD in these contexts? This comment has come up a few times in your review.

Reply:
I think this will definitely be helpful since there is a general mantra to use MUST over SHOULD where possible and to a casual reader it seems that many of the SHOULDs are closer to MUSTs.

Reply:
Will add a paragraph in META to explain why I use SHOULD for referring to test file best practices, which we can't enforce with a schema.

Reply:
This is now done (META links to Design Rationale, as that seems a better place to discuss in more detail). Please resolve after reading.